langchain/src/vectorstores/cloudflare_vectorize.ts

```diff
@@ -0,0 +1,227 @@
+    this.textKey = textKey ?? "text";
+    this.caller = new AsyncCaller({
+      maxConcurrency: 6,
+      maxRetries: 0,
+      ...asyncCallerArgs,
+    });
+  }
+
+  /**
+   * Method that adds documents to the Vectorize database.
+   * @param documents Array of documents to add.
+   * @param options Optional ids for the documents.
+   * @returns Promise that resolves with the ids of the added documents.
+   */
+  async addDocuments(
+    documents: Document[],
+    options?: { ids?: string[] } | string[]
+  ) {
+    const texts = documents.map(({ pageContent }) => pageContent);
+    return this.addVectors(
+      await this.embeddings.embedDocuments(texts),
+      documents,
+      options
+    );
+  }
+
+  /**
+   * Method that adds vectors to the Vectorize database.
+   * @param vectors Array of vectors to add.
+   * @param documents Array of documents associated with the vectors.
+   * @param options Optional ids for the vectors.
+   * @returns Promise that resolves with the ids of the added vectors.
+   */
+  async addVectors(
+    vectors: number[][],
+    documents: Document[],
+    options?: { ids?: string[] } | string[]
+  ) {
+    const ids = Array.isArray(options) ? options : options?.ids;
+    const documentIds = ids == null ? documents.map(() => uuid.v4()) : ids;
+    const vectorizeVectors = vectors.map((values, idx) => {
+      const metadata: Record<string, VectorizeVectorMetadata> = {
+        ...documents[idx].metadata,
+        [this.textKey]: documents[idx].pageContent,
+      };
+      return {
+        id: documentIds[idx],
+        metadata,
+        values,
+      };
+    });
+
+    // Stick to a limit of 500 vectors per upsert request
+    const chunkSize = 500;
+    const chunkedVectors = chunkArray(vectorizeVectors, chunkSize);
+    const batchRequests = chunkedVectors.map((chunk) =>
+      this.caller.call(async () => this.index.upsert(chunk))
+    );
+
+    await Promise.all(batchRequests);
+
+    return documentIds;
+  }
+
+  /**
+   * Method that deletes vectors from the Vectorize database.
+   * @param params Parameters for the delete operation.
+   * @returns Promise that resolves when the delete operation is complete.
+   */
+  async delete(params: VectorizeDeleteParams): Promise<void> {
+    const batchSize = 1000;
+    const batchedIds = chunkArray(params.ids, batchSize);
+    const batchRequests = batchedIds.map((batchIds) =>
+      this.caller.call(async () => this.index.deleteByIds(batchIds))
+    );
+    await Promise.all(batchRequests);
+  }
+
+  /**
+   * Method that performs a similarity search in the Vectorize database and
+   * returns the results along with their scores.
+   * @param query Query vector for the similarity search.
+   * @param k Number of top results to return.
+   * @returns Promise that resolves with an array of documents and their scores.
+   */
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number
+  ): Promise<[Document, number][]> {
```

Comment:
Would be nice to support metadata filtering at some point!

Reply:
On the roadmap!